dbms/src/AggregateFunctions/AggregateFunctionSum.cpp

```diff
@@ -29,59 +32,69 @@ using AggregateFunctionSumSimple = AggregateFunctionSum<T, typename NearestField
+
+template <typename T, typename Name>
 AggregateFunctionPtr createDecimalFunction(const IDataType * p)
 {
     if (auto dec_type = typeid_cast<const DataTypeDecimal<T> *>(p))
     {
         PrecType prec = dec_type->getPrec();
         ScaleType scale = dec_type->getScale();
-        auto [result_prec, result_scale] = SumDecimalInferer::infer(prec, scale);
+
+        auto [result_prec, result_scale] = Name::decimalInfer(prec, scale);
         auto result_type = createDecimal(result_prec, result_scale);
+
         if (checkDecimal<Decimal32>(*result_type))
-            return AggregateFunctionPtr(createWithDecimalType<AggregateFunctionSumDecimal, Decimal32>(*dec_type, prec, scale));
+            return AggregateFunctionPtr(createSumAggWithDecimalType<AggregateFunctionSumDecimal, Decimal32, Name>(*dec_type, prec, scale));
 
         if (checkDecimal<Decimal64>(*result_type))
-            return AggregateFunctionPtr(createWithDecimalType<AggregateFunctionSumDecimal, Decimal64>(*dec_type, prec, scale));
+            return AggregateFunctionPtr(createSumAggWithDecimalType<AggregateFunctionSumDecimal, Decimal64, Name>(*dec_type, prec, scale));
 
         if (checkDecimal<Decimal128>(*result_type))
-            return AggregateFunctionPtr(createWithDecimalType<AggregateFunctionSumDecimal, Decimal128>(*dec_type, prec, scale));
+            return AggregateFunctionPtr(createSumAggWithDecimalType<AggregateFunctionSumDecimal, Decimal128, Name>(*dec_type, prec, scale));
 
         if (checkDecimal<Decimal256>(*result_type))
-            return AggregateFunctionPtr(createWithDecimalType<AggregateFunctionSumDecimal, Decimal256>(*dec_type, prec, scale));
+            return AggregateFunctionPtr(createSumAggWithDecimalType<AggregateFunctionSumDecimal, Decimal256, Name>(*dec_type, prec, scale));
     }
     return nullptr;
 }
 
-template <template <typename> class Function>
+template <template <typename> class Function, typename Name>
 AggregateFunctionPtr createAggregateFunctionSum(const std::string & name, const DataTypes & argument_types, const Array & parameters)
 {
     assertNoParameters(name, parameters);
     assertUnary(name, argument_types);
 
     AggregateFunctionPtr res;
     const IDataType * p = argument_types[0].get();
-    if ((res = createDecimalFunction<Decimal32>(p)) != nullptr) {}
-    else if ((res = createDecimalFunction<Decimal64>(p)) != nullptr)
+
+    if ((res = createDecimalFunction<Decimal32, Name>(p)) != nullptr)
+    {
+    }
+    else if ((res = createDecimalFunction<Decimal64, Name>(p)) != nullptr)
     {
     }
-    else if ((res = createDecimalFunction<Decimal128>(p)) != nullptr)
+    else if ((res = createDecimalFunction<Decimal128, Name>(p)) != nullptr)
     {
     }
-    else if ((res = createDecimalFunction<Decimal256>(p)) != nullptr)
+    else if ((res = createDecimalFunction<Decimal256, Name>(p)) != nullptr)
     {
     }
     else
```

Comment:
maybe the bunch of ”if... else if“ is unnecessary. Ref to https://github.com/ClickHouse/ClickHouse/blob/master/src/AggregateFunctions/AggregateFunctionSum.cpp#L60-L61

Reply:
> maybe the bunch of ”if... else if“ is unnecessary. Ref to https://github.com/ClickHouse/ClickHouse/blob/master/src/AggregateFunctions/AggregateFunctionSum.cpp#L60-L61

Not only if it's decimal type we need to ensure, but also which precision of decimal we need to judge. So we can't just use isDecimal() to replace these if...else